[FIX] web: snap planned_date to closest choice

pivi-odoo opened the pull request:
## Current behaviour
In Field Services, when selecting a task and applying let's say the ending time of 15:34, when the date picker shows up the minutes were set to 0, because 34 is not any multiple of the `timePickerIncrement` (which is 5 by default).

## Expected behaviour
Even if the minutes are not a multiple of the `timePickerIncrement`, we should round to the closest one.

## Steps to reproduce
- Install Field Services
- Fields Services > Planning > By Project > Day
- Choose a task, and play around with the minutes of either starting and ending (you need to save each time to see the effect)
- If the minutes are a multiple of `timePickerIncrement` = 5, then there is the correct selection, but for everything else it is 0.

## Reason for the problem
When rendering the date range picker, only if the minute is exactly a multiple of `timePickerIncrement` and only then it is selected. For all other cases nothing is selected, therefor the first item is chosen by default.

## Fix
We select the closest multiple of `timePickerIncrement` by truncation. Doing a regular rounding would require too much change (handling rollover X:59 rounded to X+1:00). Therefor truncation was chosen.

## Affected versions
- 14.0
- 15.0
- saas-15.2
- 16.0
- saas-16.1
- master
---
opw-3103490

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

nle-odoo reviewed (approved):
looks good to me.

nle-odoo commented on addons/web/static/lib/daterangepicker/daterangepicker.js:
```diff
@@ -942,7 +942,7 @@
                     disabled = true;
 
-                if (selected.minute() == i && !disabled) {
+                if (Math.floor(selected.minute() / this.timePickerIncrement) * this.timePickerIncrement === i && !disabled) {
```
these math operation should be quick, but maybe `Math.floor(selected.minute() / this.timePickerIncrement) * this.timePickerIncrement` could be computed as a `const` before the loop.

pivi-odoo replied:
Thanks for the review.
Yeah I guess it is possible, but IMHO I don't think it's worth it.
This is not a hot code path and worse case scenario the `timePickerIncrement` is 1, so it will be a loop of 60 iterations which is nothing. In the end it's just a `div` (the floor is quasi free, because it's just taking the quotient part) with a `mul`. Replacing it with an out-of-loop const named let's say `truncatedMinutes` will force the next reader to scroll up just to see what is the formula, instead of having it inlined.
Also I am pretty sure the JIT that will be executing this code will surely have some Loop-invariant code motion optimization enabled. 
But all of this is just my opinion, if the JS team says they think it is necessary, yeah sure I will hoist the formula out of the loop.

pivi-odoo commented:
The lib hasn't been updated in the last 2 years.

pivi-odoo commented:
@aab-odoo 👋🏼 I've added the odoo patch as requested :)

pivi-odoo commented:
Hi @aab-odoo 👋🏼,
I've updated the code, it should reads from the input fields now.
Since using the API `setStartDate` and `setEndDate` does the proper rounding with rollover, I've removed the patch from the lib/daterangepicker.js

aab-odoo commented on addons/web/static/src/js/fields/basic_fields.js:
```diff
@@ -785,16 +785,23 @@ var FieldDateRange = InputField.extend({
             }
         };
         window.addEventListener('scroll', this._onScroll, true);
+        if (this.recordData.planned_date_begin && this.recordData.planned_date_end) {
```
this is a framework component... business field names should never been hardcoded here

pivi-odoo replied:
My bad, I've updated the code to be generic.

pivi-odoo commented:
Tests are good now.

mcm-odoo commented:
Hello @pivi-odoo 
I tested your last build in event app and it crashed when I clicked on the date range field.
Also, your issue looks like the one in this PR https://github.com/odoo/odoo/pull/107508

pivi-odoo commented:
Hello @mcm-odoo,
I've fix the stacktrace, but oddly enough I've stumbled upon another problem.
Let's say we edit both the start and end date by hand by typing in the input field, we close the datepicker, then we click on top of the startDate input, the record for the endDate is not being updated with the value from the input field. (but if you click on the endDate input field instead, the record contains the correct date in both entries from startDate and endDate). We've looked into it with Hubert, but without much luck, if you have any suggestion?

mcm-odoo commented:
@robodoo r+
